zebra-state/src/in_memory/chains.rs

```diff
@@ -3,18 +3,98 @@ use crate::{Request, Response};
 use futures::prelude::*;
 use std::{
     pin::Pin,
+    sync::Arc,
     task::{Context, Poll},
 };
-use tower::Service;
+use tower::{Service, ServiceExt};
+use zebra_chain::{
+    block::{Block, BlockHeaderHash},
+    types::BlockHeight,
+};
+
+/// A representation of the chain state at a given height
+#[derive(Clone, Debug)]
+struct ChainState {
+    block: Arc<Block>,
+    // TODO: add more state here
+}
+
+/// A persistent data structure representing the end of a chain
+struct VolatileChain(im::OrdMap<BlockHeight, ChainState>);
+
+impl VolatileChain {
+    fn contains(&self, height: BlockHeight) -> bool {
+        todo!()
+    }
+}
 
 /// A service wrapper that tracks multiple chains, handles reorgs, and persists
 /// blocks to disk once they're past the reorg limit
 pub(crate) struct ChainsState<S> {
     /// The inner state service that only tracks a single chain
     inner: S,
+    /// The set of chains
+    //
+    // might need to use a map type and pop / reinsert with cummulative work as the index
+    chains: Vec<VolatileChain>,
+}
+
+impl<S> ChainsState<S> {
+    fn insert(&mut self, block: impl Into<Arc<Block>>) -> Result<BlockHeaderHash, Error> {
+        let block = block.into();
+        let hash = block.hash();
+        let height = block.coinbase_height().unwrap();
+        let parent_height = BlockHeight(height.0 - 1);
+
+        for chain in self
+            .chains
+            .iter_mut()
+            .filter(|chain| chain.contains(parent_height))
+        {
+            let parent_state = chain
+                .0
+                .get(&parent_height)
+                .expect("block with one less height must exist");
```

Comment:
I don't think this depends on the map being ordered, we can't panic here because we filtered out chains that don't contain an element at `parent_height`, and then we retrieve that element. Come to think of it I can re-write this in a way that doesn't even panic, though I bet in release mode rust optimizes out the panic as is.